java/server/src/org/openqa/grid/common/SeleniumProtocol.java

```diff
@@ -17,8 +17,44 @@
 
 package org.openqa.grid.common;
 
+import static org.openqa.grid.common.RegistrationRequest.PATH;
+import static org.openqa.grid.common.RegistrationRequest.SELENIUM_PROTOCOL;
+
+import org.openqa.grid.common.exception.GridException;
+import org.openqa.selenium.remote.DesiredCapabilities;
+
+import java.util.Arrays;
+import java.util.Map;
+
 public enum SeleniumProtocol {
-  Selenium, WebDriver;
+  Selenium("/selenium-server/driver"),
+  WebDriver("/wd/hub");
+  private String path;
+
+  SeleniumProtocol(String path) {
+    this.path = path;
+  }
+
+  public static SeleniumProtocol getProtocol(DesiredCapabilities capability) {
+    String type = (String) capability.getCapability(SELENIUM_PROTOCOL);
+    if (type == null || type.trim().isEmpty()) {
+      return WebDriver;
+    }
+    try {
+      return SeleniumProtocol.valueOf(type);
+    } catch (IllegalArgumentException e) {
+      throw new GridException(type + " isn't a valid protocol type for grid. Valid values :[" +
+                              Arrays.toString(values()) + "]", e);
+    }
+  }
+
+  public String getPath(Map<String, Object> capabilities) {
```

Comment:
Fixed. We now work with only the `Map` since `DesiredCapabilities` is not available for access from within `TestSlot` (that's the place wherein `getPathConsideringDesiredCapabilities()` would be accessed).